### PR TITLE
Add InsertionResult to LocalSlamResult.

### PR DESCRIPTION
**Pull request by cschuet**
As discussed on https://github.com/googlecartographer/cartographer/pull/574 I need to include the `InsertionResult` in the `LocalSlamResult` in order to add a `Node` in the `PoseGraph` in the cloud Cartographer instance.

### CODE REVIEW

File: cartographer/internal/mapping/global_trajectory_builder.h [hunk 1]
```diff
@@ -59,17 +59,26 @@ class GlobalTrajectoryBuilder : public mapping::TrajectoryBuilderInterface {
...
     if (matching_result->insertion_result != nullptr) {
       node_id = ::cartographer::common::make_unique<mapping::NodeId>(
           pose_graph_->AddNode(
               matching_result->insertion_result->constant_data, trajectory_id_,
               matching_result->insertion_result->insertion_submaps));
       CHECK_EQ(node_id->trajectory_id, trajectory_id_);
+      insertion_result = common::make_unique<InsertionResult>();
+      insertion_result->constant_data =
+          matching_result->insertion_result->constant_data;
+      for (const auto& insertion_submap :
```
Comment (wohe) [inline cartographer/internal/mapping/global_trajectory_builder.h]: Why not just
```
insertion_result->insertion_submaps = matching_result->insertion_result->insertion_submaps
```
or
```
common::make_unique<InsertionResult>(InsertionResult{
  matching_result->insertion_result->constant_data;
  matching_result->insertion_result->insertion_submaps;
});
```

File: cartographer/mapping/trajectory_builder_interface.h [hunk 2]
```diff
@@ -44,14 +44,20 @@ proto::TrajectoryBuilderOptions CreateTrajectoryBuilderOptions(
 // optimized pose estimates.
 class TrajectoryBuilderInterface {
  public:
+  struct InsertionResult {
+    std::shared_ptr<const mapping::TrajectoryNode::Data> constant_data;
+    std::vector<std::shared_ptr<const Submap>> insertion_submaps;
+  };
```
Comment (wohe) [inline cartographer/mapping/trajectory_builder_interface.h]: Wouldn't it make sense to include the `NodeId` here as well and just pass one struct?

File: cartographer/internal/mapping/global_trajectory_builder.h (hunk 1 above)
Comment (cschuet) [reply cartographer/internal/mapping/global_trajectory_builder.h]: That does not compile.

Note that the `matching_result->insertion_result->insertion_submaps` are of type `std::vector<mapping_2d/3d::Submap>` while the newly introduced one's are `std::vector<mapping::Submap>`, i.e. the base class.

File: cartographer/mapping/trajectory_builder_interface.h (hunk 2 above)
Comment (cschuet) [reply cartographer/mapping/trajectory_builder_interface.h]: Is it true that you have a `NodeId` iff you have an `InsertionResult`.

Comment (wohe) [reply cartographer/mapping/trajectory_builder_interface.h]: I think so. The comment below also says that.

File: cartographer/internal/mapping/global_trajectory_builder.h (hunk 1 above)
Comment (wohe) [reply cartographer/internal/mapping/global_trajectory_builder.h]: I see. What about the range constructor then, i.e.
```
common::make_unique<InsertionResult>(InsertionResult{
  matching_result->insertion_result->constant_data,
  std::vector<std::shared_ptr<const Submap>>(
    matching_result->insertion_result->insertion_submaps.begin(),
    matching_result->insertion_result->insertion_submaps.end())});
```
?

File: cartographer/internal/mapping/global_trajectory_builder.h [hunk 3]
```diff
@@ -59,17 +59,24 @@ class GlobalTrajectoryBuilder : public mapping::TrajectoryBuilderInterface {
       return;
     }
     std::unique_ptr<mapping::NodeId> node_id;
```
Comment (wohe) [inline cartographer/internal/mapping/global_trajectory_builder.h]: Remove?

File: cartographer/internal/mapping/global_trajectory_builder.h [hunk 4]
```diff
@@ -59,17 +59,24 @@ class GlobalTrajectoryBuilder : public mapping::TrajectoryBuilderInterface {
...
+    std::unique_ptr<InsertionResult> insertion_result;
     if (matching_result->insertion_result != nullptr) {
       node_id = ::cartographer::common::make_unique<mapping::NodeId>(
```
Comment (wohe) [inline cartographer/internal/mapping/global_trajectory_builder.h]: Remove the `unique_ptr`which is no longer used?

File: cartographer/internal/mapping/global_trajectory_builder.h (hunk 3 above)
Comment (cschuet) [reply cartographer/internal/mapping/global_trajectory_builder.h]: Thanks!

File: cartographer/internal/mapping/global_trajectory_builder.h (hunk 4 above)
Comment (cschuet) [reply cartographer/internal/mapping/global_trajectory_builder.h]: Done.

**Comment by cschuet** [comment]
@wally-the-cartographer merge
@wohe Thanks for the review!

**Pull request merged**